### PR TITLE
fix maas juju deploy --to lxd

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Remove code duplicated from state.AssignUnitWithPlacement in BridgePolicy determineContainerSpaces().  Cause failure to deploy to a container on a MAAS machine due to different parsing of application EndpointBindings than in state.

## QA steps
On lxd, maas, aws and any others:
1. juju deploy
1. juju deploy --to lxd

On maas
1. juju deploy --bind endpoint=space

### CODE REVIEW

File: network/containerizer/bridgepolicy_integration_test.go [hunk 1]
```diff
@@ -20,7 +15,9 @@ import (
 	"github.com/juju/juju/network/containerizer"
 	"github.com/juju/juju/state"
 	statetesting "github.com/juju/juju/state/testing"
-	"github.com/juju/juju/testcharms"
+	jc "github.com/juju/testing/checkers"
```
Comment (hmlanigan) [inline network/containerizer/bridgepolicy_integration_test.go]: fixing this.

**Comment by hmlanigan** [comment]
Working on more changes to the container and provisioner interfaces to pull apart some of the dependencies for testing etc.  Will be in another PR.

File: network/containerizer/bridgepolicy.go [hunk 2]
```diff
@@ -32,8 +32,10 @@ var skippedDeviceNames = set.NewStrings(
 // for guests inside of a host machine, along with the creation of network
 // devices on those bridges for the containers to use.
 type BridgePolicy struct {
-	// spaces is a cache of the model's spaces.
+	// spaces is a cache of the model's space IDs by name.
```
Comment (howbazaar) [inline network/containerizer/bridgepolicy.go]: A map of space name to what?

File: network/containerizer/bridgepolicy.go [hunk 3]
```diff
@@ -32,8 +32,10 @@ var skippedDeviceNames = set.NewStrings(
...
 	spaces map[string]string
+	// spaces is a cache of the model's space names by ID.
+	spaceNames map[string]string
```
Comment (howbazaar) [inline network/containerizer/bridgepolicy.go]: If this is a map of names by ID, what are we mapping to? Are we mapping to the name? Perhaps we should rename the other one to be `spaceIDs`?

File: network/containerizer/bridgepolicy.go [hunk 4]
```diff
@@ -393,6 +367,25 @@ func (p *BridgePolicy) determineContainerSpaces(
...
+// spaceNamesForPrinting return a sorted, comma delimited, string containing
+// the space names for each of the given space ids.
+func (p *BridgePolicy) spaceNamesForPrinting(ids set.Strings) string {
+	if ids.Size() == 0 {
+		return "<none>"
+	}
+	names := set.NewStrings()
+	for _, id := range ids.Values() {
+		if name, ok := p.spaceNames[id]; ok {
+			names.Add("\"" + name + "\"")
```
Comment (howbazaar) [inline network/containerizer/bridgepolicy.go]: Would `names.Add(fmt.Sprintf("%q", name))` be easier?

File: network/containerizer/bridgepolicy.go (hunk 3 above)
Comment (howbazaar) [reply network/containerizer/bridgepolicy.go]: I clearly can't read properly. space ID -> space name is clearly mentioned.

File: network/containerizer/bridgepolicy.go [hunk 5]
```diff
@@ -58,8 +60,14 @@ func NewBridgePolicy(cfgGetter environs.ConfigGetter, st SpaceBacking) (*BridgeP
 		return nil, errors.Annotate(err, "getting space name to space ID map")
 	}
 
+	spacesNames, err := st.SpaceNamesByID()
```
Comment (manadart) [inline network/containerizer/bridgepolicy.go]: We are hitting the DB multiple times to get the same data in a different "shape" at least here, but I seem to recall in other places too.

Given the emerging ubiquity of space lookups, and the fact that we have 3 state package methods for the purpose, I think we should do the following:
- Remove `SpaceNamesByID`, `SpaceIDsByName`, `SpaceInfosByID`.
- Instead have a single method to return all spaces as `network.SpaceInfos`.
- Move the lookup generations to the `SpaceInfos` type.

Then we only ever hit the DB once to materialise a the set of spaces, and the reshaping for lookups is done in-memory.

File: network/containerizer/bridgepolicy.go [hunk 6]
```diff
@@ -393,6 +367,25 @@ func (p *BridgePolicy) determineContainerSpaces(
 	return spaces.Values(), nil
 }
 
+// spaceNamesForPrinting return a sorted, comma delimited, string containing
+// the space names for each of the given space ids.
+func (p *BridgePolicy) spaceNamesForPrinting(ids set.Strings) string {
```
Comment (manadart) [inline network/containerizer/bridgepolicy.go]: Here again, if we had a single member of type `SpaceInfos`, we could implement a filter on that type and just lean on its `String` method.

File: network/containerizer/bridgepolicy.go (hunk 4 above)
Comment (hmlanigan) [reply network/containerizer/bridgepolicy.go]: yes, updated

File: network/containerizer/bridgepolicy.go (hunk 3 above)
Comment (hmlanigan) [reply network/containerizer/bridgepolicy.go]: I renamed spaces to spaceIDs and updated the comments which both said spaces.

File: network/containerizer/bridgepolicy.go (hunk 5 above)
Comment (hmlanigan) [reply network/containerizer/bridgepolicy.go]: Will take a look at this for a follow on.  Here an in other places.

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**